pkg/challenge/challenger.go

```diff
@@ -18,16 +19,105 @@ type OuterChallenge struct {
 	Msg []byte `msgpack:"msg"`
 }
 
+func (o *OuterChallenge) Verify(counterParty ecdsa.PublicKey) error {
+	return echelper.VerifySignature(counterParty, o.Msg, o.Sig)
+}
+
+func (o *OuterChallenge) RequestData() ([]byte, error) {
+	inner, err := o.inner()
+	if err != nil {
+		return nil, err
+	}
+	return inner.RequestData, nil
+}
+
+func (o *OuterChallenge) Respond(signer crypto.Signer, responseData []byte) ([]byte, error) {
+	innerChallenge, err := o.inner()
+	if err != nil {
+		return nil, err
+	}
+
+	pubSigningKeyPem, err := echelper.PublicEcdsaKeyToPem(signer.Public().(*ecdsa.PublicKey))
+	if err != nil {
+		return nil, fmt.Errorf("marshalling public signing key to pem: %w", err)
+	}
+
+	innerResponse, err := msgpack.Marshal(InnerResponse{
+		PublicSigningKey: pubSigningKeyPem,
+		ChallengeData:    innerChallenge.ChallengeData,
+		ResponseData:     responseData,
+		TimeStamp:        time.Now().Unix(),
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("marshaling inner box: %w", err)
+	}
+
+	signature, err := echelper.Sign(signer, innerResponse)
+	if err != nil {
+		return nil, fmt.Errorf("signing challenge: %w", err)
+	}
+
+	sealed, pub, err := echelper.SealNaCl(innerResponse, &innerChallenge.PublicEncryptionKey)
+	if err != nil {
+		return nil, fmt.Errorf("sealing challenge inner box: %w", err)
+	}
+
+	outerResponse := &OuterResponse{
+		PublicEncryptionKey: *pub,
+		Sig:                 signature,
+		Msg:                 sealed,
+		ChallengeId:         innerChallenge.ChallengeId,
+	}
+
+	return msgpack.Marshal(outerResponse)
+}
+
+func (o *OuterChallenge) inner() (*InnerChallenge, error) {
+	var inner InnerChallenge
+	if err := msgpack.Unmarshal(o.Msg, &inner); err != nil {
+		return nil, fmt.Errorf("unmarshaling inner challenge: %w", err)
+	}
+	return &inner, nil
+}
+
+func (o *OuterChallenge) RespondPng(signer crypto.Signer, responseData []byte) ([]byte, error) {
+	response, err := o.Respond(signer, responseData)
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	if err := krypto.ToPng(&buf, response); err != nil {
+		return nil, fmt.Errorf("encoding data to png: %w", err)
+	}
+
+	return buf.Bytes(), nil
+}
+
+func UnmarshalOuterChallenge(outerChallengeBytes []byte) (*OuterChallenge, error) {
+	var outerChallenge OuterChallenge
+	if err := msgpack.Unmarshal(outerChallengeBytes, &outerChallenge); err != nil {
+		return nil, err
+	}
+	return &outerChallenge, nil
+}
+
+func MarshalOuterChallenge(outerChallenge *OuterChallenge) ([]byte, error) {
+	return msgpack.Marshal(outerChallenge)
+}
```

Comment:
Do you think this should me a method on outerChallenge directly?

Reply:
done